File: 백준/Gold/20366. 같이 눈사람 만들래？/같이 눈사람 만들래？.java
import static java.lang.Math.abs;
import static java.lang.Math.min;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int N;
    static int[] snow;
    static Snow[] snowMan;
    static long ans;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());

        snow = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            snow[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(snow);

        int N2 = N * (N - 1) / 2;
        snowMan = new Snow[N2];

        int idx = 0;
        for (int i = 0; i < N; i++) {
            for (int j = i + 1; j < N; j++) {
                snowMan[idx++] = new Snow(snow[i] + snow[j], i, j);
            }
        }

        Arrays.sort(snowMan);
        ans = 1000000001;
        idx = -1;
        while (++idx < N2 - 1) {
            if (snowMan[idx].contains(snowMan[idx + 1])) {
                continue;
            }
            ans = min(ans, abs(snowMan[idx].height - snowMan[idx + 1].height));
        }
        System.out.println(ans);
    }

    static class Snow implements Comparable{
        public int height, snow1, snow2;

        public Snow(int height, int snow1, int snow2) {
            this.height = height;
            this.snow1 = snow1;
            this.snow2 = snow2;
        }

        boolean contains(Snow snow) {
            if (snow.snow1 == this.snow1 || snow.snow1 == this.snow2 || snow.snow2 == this.snow1 || snow.snow2 == this.snow2) return true;
            return false;
        }

        @Override
        public int compareTo(Object o) {
            Snow snow = (Snow) o;
            if (height < snow.height) {
                return -1;
            } else if (height > snow.height) {
                return 1;
            } else {
                return 0;
            }
        }
    }
}

File: 백준/Gold/20366. 같이 눈사람 만들래？/같이 눈사람 만들래？.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// 1. 4개를 고른다
// 2. 2개씩 짝지어 눈사람을 만드는데, 키 차이가 작을 수록 좋다.
// 3. 아래에 더 큰게 가야한다. (중요?)
// 총 600개의 눈덩이가 있다.
class SnowMan {

public:
    long long snow[2];
    long long length;
    SnowMan(long long snow1, long long snow2, long long length) {
        this->snow[0] = snow1;
        this->snow[1] = snow2;
        this->length = length;
    }
};

bool compareSnow(const SnowMan &snowMan1, const SnowMan &snowMan2) {
    return snowMan1.length < snowMan2.length;
}

int main() {

    int N;
    cin >> N;
    vector<long long> snows;
    for (int i = 0; i < N; i++) {
//        cin >> snows[i];
        long long input;
        cin >> input;
        snows.push_back(input);
    }

    sort(snows.begin(), snows.end());

    vector<SnowMan> snowMans;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (i == j) continue;
            SnowMan snowMan = SnowMan(i, j, snows[i] + snows[j]);
            snowMans.emplace_back(snowMan);
        }
    }

    sort(snowMans.begin(), snowMans.end(), compareSnow);

    long long result = 1000000001;
    int index = -1;
    while (++index < snowMans.size() - 1) {
        SnowMan snowMan1 = snowMans[index];
        SnowMan snowMan2 = snowMans[index + 1];
        if (snowMan1.snow[0] == snowMan2.snow[0] || snowMan1.snow[0] == snowMan2.snow[1]
            || snowMan1.snow[1] == snowMan2.snow[0] || snowMan1.snow[1] == snowMan2.snow[1]) continue;
        result = min(result, abs(snowMan1.length - snowMan2.length));
    }

    cout << result;
}


// 2 3 5 5 9

// 전체에서 2개 빼고 2개 선택 -> 598C2 -> 200 * 587 -> 대충 12000개
// 다 곱하면 안돼